examples/IBLevelSet/ex6/example.cpp

```diff
@@ -0,0 +1,627 @@
+                           << std::endl;
+            }
+
+            // At specified intervals, write visualization and restart files,
+            // and print out timer data.
+            iteration_num += 1;
+            const bool last_step = !navier_stokes_integrator->stepsRemaining();
+            if (dump_viz_data && uses_visit && (iteration_num % viz_dump_interval == 0 || last_step))
+            {
+                pout << "Writing visualization files...\n\n";
+                navier_stokes_integrator->setupPlotData();
+                visit_data_writer->writePlotData(patch_hierarchy, iteration_num, loop_time);
+            }
+            if (dump_restart_data && (iteration_num % restart_dump_interval == 0 || last_step))
+            {
+                pout << "Writing restart files...\n\nn";
+                RestartManager::getManager()->writeRestartFile(restart_dump_dirname, iteration_num);
+            }
+            if (dump_timer_data && (iteration_num % timer_dump_interval == 0 || last_step))
+            {
+                pout << "Writing timer data...\n\n";
+                TimerManager::getManager()->print(plog);
+            }
+
+            if (SAMRAI_MPI::getRank() == 0)
+            {
+                const Eigen::Vector3d& rbd_posn = bp_rbd->getCurrentCOMPosn();
+                const Eigen::Vector3d& rbd_trans_vel = bp_rbd->getCurrentCOMTransVelocity();
+
+                Eigen::Vector3d hydro_force_pressure, hydro_force_viscous, hydro_torque_pressure, hydro_torque_viscous;
+                bp_rbd->getHydrodynamicForceTorque(
+                    hydro_force_pressure, hydro_force_viscous, hydro_torque_pressure, hydro_torque_viscous);
+
+                rbd_stream.precision(12);
+                rbd_stream.setf(ios::fixed, ios::floatfield);
+                rbd_stream << loop_time << "\t" << rbd_posn[1] << "\t" << rbd_trans_vel[1] << std::endl;
+
+                ft_stream.precision(12);
+                ft_stream.setf(ios::fixed, ios::floatfield);
+                ft_stream << loop_time << "\t" << hydro_force_pressure[0] << "\t" << hydro_force_pressure[1] << "\t"
+                          << hydro_force_pressure[2] << "\t" << hydro_force_viscous[0] << "\t" << hydro_force_viscous[1]
+                          << "\t" << hydro_force_viscous[2] << "\t" << hydro_torque_pressure[0] << "\t"
+                          << hydro_torque_pressure[1] << "\t" << hydro_torque_pressure[2] << "\t"
+                          << hydro_torque_viscous[0] << "\t" << hydro_torque_viscous[1] << "\t"
+                          << hydro_torque_viscous[2] << std::endl;
+            }
+        }
+
+        // Close the logging streams.
+        if (SAMRAI_MPI::getRank() == 0)
+        {
+            vol_stream.close();
+            rbd_stream.close();
+            ft_stream.close();
+        }
+
+        // Delete dumb pointers.
```

Comment:
Could we change "dumb" pointers to `std::unique_ptr`?

Reply:
This requires changing the interface of many hierarchy integrators, and none of the examples do that. Can you please point to an example if you know any that does this?

Reply:
It shouldn't require changing any interfaces --- `unique_ptr<T>::get()` will return the underlying raw pointer `T*`.

(I think that we should try to move towards an API in which we use raw pointers or references to pass objects without passing ownership of the objects.)

Reply:
I have made the requested changes. Please take a look and let me know.